ykrt/src/location.rs

```diff
@@ -334,9 +321,25 @@ impl Drop for Location {
         let ls = self.load(Ordering::Relaxed);
         if !ls.is_counting() {
             debug_assert!(!ls.is_locked());
-            let hr = unsafe { ls.hot_location() };
-            unsafe {
-                Box::from_raw(hr);
+            self.lock().unwrap();
+            let hl = unsafe { ls.hot_location() };
+            if let HotLocation::Compiled(_) = hl {
+                // FIXME: we can't drop this memory as another thread may still be executing the
+                // trace that's pointed to. There should be a ref count that we decrement and free
+                // memory if it reaches zero.
+                self.unlock();
+            } else if let HotLocation::Compiling = hl {
+                // Another thread has a pointer to the underlying `HotLocation`, so we have to
+                // signal to that thread that they need to free the `HotLocation`.
+                *hl = HotLocation::Dropped;
```

Comment:
I take it this scheme is devised so as to avoid using a `RefCell` (which would be a simpler, but less-performant way of doing things).

When would the user drop a `Location`, besides when the interpreter is ending? Is it worth the added complexity?

Reply:
`RefCell`s aren't `Sync` so you'd need to add a `Mutex` (or similar) around them.

Reply:
Uh. Did I mean to say `Arc<RefCell>`?

Basically, why add this complexity if the only time a location would be dropped is a time where you no longer care about performance?

Reply:
> Did I mean to say Arc<RefCell>?

Only you know what you meant to say :p

> Basically, why add this complexity if the only time a location would be dropped is a time where you no longer care about performance?

If I was to put back an `Arc` or `Mutex`, we'd end up with the intermediate state where a `Location` has compiled but isn't yet in the `Compiling` state. In a sense this whole PR is about getting rid of that.